File: src/main/java/org/dukcode/ps/codetree/trail2/lesson2/Intro369GamesUsingFunctions.java
package org.dukcode.ps.codetree.trail2.lesson2;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.StringTokenizer;

/**
 * @see <a
 * href="https://www.codetree.ai/ko/trails/complete/curated-cards/intro-369-games-using-functions/description">LINK</a>
 */
public class Intro369GamesUsingFunctions {

  private static BufferedReader br;
  private static BufferedWriter bw;
  private static StringTokenizer st;

  public static void main(String[] args) throws IOException {
    br = new BufferedReader(new InputStreamReader(System.in));
    bw = new BufferedWriter(new OutputStreamWriter(System.out));

    st = new StringTokenizer(br.readLine());
    int a = Integer.parseInt(st.nextToken());
    int b = Integer.parseInt(st.nextToken());

    bw.write(String.valueOf(solve(a, b)));

    br.close();
    bw.close();

  }

  private static int solve(int a, int b) {
    if (a > b) {
      return solve(b, a);
    }

    int cnt = 0;
    for (int num = a; num <= b; num++) {
      if (check(num)) {
        cnt++;
      }
    }

    return cnt;
  }

  private static boolean check(int num) {
    if (num % 3 == 0) {
      return true;
    }

    while (num > 0) {
      int digit = num % 10;

      if (digit == 3 || digit == 6 || digit == 9) {
        return true;
      }

      num /= 10;
    }

    return false;
  }
}

File: src/main/kotlin/org/dukcode/ps/codetree/trail02/chapter01/lesson02/Intro369GamesUsingFunctions.kt
package org.dukcode.ps.codetree.trail02.chapter01.lesson02

import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.util.StringTokenizer

private val br = BufferedReader(InputStreamReader(System.`in`))
private val bw = BufferedWriter(OutputStreamWriter(System.out))

/**
 * @see <a href="https://www.codetree.ai/ko/trails/complete/curated-cards/intro-369-games-using-functions/description">LINK</a>
 */
fun main() {
    val st = StringTokenizer(br.readLine())

    val a = st.nextToken().toInt()
    val b = st.nextToken().toInt()

    bw.write(count369(a, b).toString())

    bw.flush()
}

private fun count369(
    frIn: Int,
    toIn: Int,
): Int {
    var ret = 0
    for (num in frIn..toIn) {
        if (has369(num) || num % 3 == 0) {
            ret++
        }
    }

    return ret
}

fun has369(num: Int): Boolean {
    var value = num
    while (value > 0) {
        val digit = value % 10

        if (digit == 3 || digit == 6 || digit == 9) {
            return true
        }

        value /= 10
    }

    return false
}
